sklearn/tree/_partitioner.pyx

```diff
@@ -110,102 +126,85 @@ cdef class DensePartitioner:
+            if isnan(current_feature_value):
+                n_missing += 1
+            elif current_feature_value < min_feature_value:
+                min_feature_value = current_feature_value
+            elif current_feature_value > max_feature_value:
+                max_feature_value = current_feature_value
 
         min_feature_value_out[0] = min_feature_value
         max_feature_value_out[0] = max_feature_value
         self.n_missing = n_missing
 
     cdef inline void next_p(self, intp_t* p_prev, intp_t* p) noexcept nogil:
-        """Compute the next p_prev and p for iterating over feature values.
-
-        The missing values are not included when iterating through the feature values.
         """
-        cdef intp_t end_non_missing = self.end - self.n_missing
+        Compute the next p_prev and p for iterating over feature values.
 
-        while (
-            p[0] + 1 < end_non_missing and
-            self.feature_values[p[0] + 1] <= self.feature_values[p[0]] + FEATURE_THRESHOLD
-        ):
+        - if self.missing_on_the_left: go over the p in [start + n_missing + 1, end[
+        - else: go over the p in [start, end_non_missing]
+            when p=end_non_missing, this means all non-missing values go to the left
+            and all missing to the right
+        """
+        cdef intp_t end_non_missing = (
+            self.end if self.missing_on_the_left
+            else self.end - self.n_missing)
+
+        if p[0] == end_non_missing and not self.missing_on_the_left:
+            # skip the missing values up to the end
+            # (which will end the for loop in the best split function)
+            p[0] = self.end
+            p_prev[0] = self.end
+        else:
+            if self.missing_on_the_left and p[0] == self.start:
+                # skip the missing values up to the first non-missing value:
+                p[0] = self.start + self.n_missing
             p[0] += 1
-
-        p_prev[0] = p[0]
-
-        # By adding 1, we have
-        # (feature_values[p] >= end) or (feature_values[p] > feature_values[p - 1])
-        p[0] += 1
+            while (
+                p[0] < end_non_missing and
+                self.feature_values[p[0]] <= self.feature_values[p[0] - 1] + FEATURE_THRESHOLD
+            ):
+                p[0] += 1
+            p_prev[0] = p[0] - 1
 
     cdef inline intp_t partition_samples(
         self,
-        float64_t current_threshold
+        float64_t current_threshold,
+        bint missing_go_to_left
     ) noexcept nogil:
         """Partition samples for feature_values at the current_threshold."""
         cdef:
             intp_t p = self.start
-            intp_t partition_end = self.end - self.n_missing
-            intp_t[::1] samples = self.samples
-            float32_t[::1] feature_values = self.feature_values
+            intp_t partition_end = self.end
+            intp_t* samples = &self.samples[0]
+            float32_t* feature_values = &self.feature_values[0]
```

Comment:
Why do we need the pointers? Can we just pass to swap the references of self.features_values[0] and self.samples[0]

Or does that not work? It's been awhile since I've thought about the implications in cython :p